pkg/cloudprovider/providers/gce/cloud/mock/mock.go

```diff
@@ -74,3 +82,115 @@ func RemoveInstanceHook(ctx context.Context, key *meta.Key, req *ga.TargetPoolsR
 
 	return nil
 }
+
+func convertAndInsertAlphaForwardingRule(key *meta.Key, obj gceObject, mRules map[meta.Key]*cloud.MockForwardingRulesObj, version meta.Version) (bool, error) {
+	if !key.Valid() {
+		return false, fmt.Errorf("invalid GCE key (%+v)", key)
+	}
+
+	if _, ok := mRules[*key]; ok {
+		err := &googleapi.Error{
+			Code:    http.StatusConflict,
+			Message: fmt.Sprintf("MockForwardingRule %v exists", key),
+		}
+		return false, err
+	}
+
+	enc, err := obj.MarshalJSON()
+	if err != nil {
+		return false, err
+	}
+	var fwdRule alpha.ForwardingRule
+	if err := json.Unmarshal(enc, &fwdRule); err != nil {
+		return false, err
+	}
+	// Set the default values for the Alpha fields.
+	if fwdRule.NetworkTier == "" {
+		fwdRule.NetworkTier = cloud.NetworkTierDefault.ToGCEValue()
+	}
+
+	fwdRule.Name = key.Name
+	if fwdRule.SelfLink == "" {
+		fwdRule.SelfLink = cloud.SelfLink(version, "mock-project", "forwardingRules", key)
+	}
+
+	mRules[*key] = &cloud.MockForwardingRulesObj{Obj: fwdRule}
+	return true, nil
+}
+
+// InsertFwdRuleHook mocks inserting a ForwardingRule. ForwardingRules are
+// expected to default to Premium tier if no NetworkTier is specified.
+func InsertFwdRuleHook(ctx context.Context, key *meta.Key, obj *ga.ForwardingRule, m *cloud.MockForwardingRules) (bool, error) {
+	return convertAndInsertAlphaForwardingRule(key, obj, m.Objects, meta.VersionGA)
+}
+
+// InsertBetaFwdRuleHook mocks inserting a BetaForwardingRule.
+func InsertBetaFwdRuleHook(ctx context.Context, key *meta.Key, obj *beta.ForwardingRule, m *cloud.MockForwardingRules) (bool, error) {
+	return convertAndInsertAlphaForwardingRule(key, obj, m.Objects, meta.VersionBeta)
+}
+
+// InsertAlphaFwdRuleHook mocks inserting an AlphaForwardingRule.
+func InsertAlphaFwdRuleHook(ctx context.Context, key *meta.Key, obj *alpha.ForwardingRule, m *cloud.MockForwardingRules) (bool, error) {
+	return convertAndInsertAlphaForwardingRule(key, obj, m.Objects, meta.VersionAlpha)
+}
+
+// Used to assign Addresses with no IP a unique IP address
+var ipCounter = 1
+
+func convertAndInsertAlphaAddress(key *meta.Key, obj gceObject, mAddrs map[meta.Key]*cloud.MockAddressesObj, version meta.Version) (bool, error) {
+	if !key.Valid() {
+		return false, fmt.Errorf("invalid GCE key (%+v)", key)
+	}
+
+	if _, ok := mAddrs[*key]; ok {
+		err := &googleapi.Error{
+			Code:    http.StatusConflict,
+			Message: fmt.Sprintf("MockAddresses %v exists", key),
+		}
+		return false, err
+	}
+
+	enc, err := obj.MarshalJSON()
+	if err != nil {
+		return false, err
+	}
+	var addr alpha.Address
+	if err := json.Unmarshal(enc, &addr); err != nil {
+		return false, err
+	}
+
+	// Set default values used in tests
+	addr.Name = key.Name
+	if addr.SelfLink == "" {
+		addr.SelfLink = cloud.SelfLink(version, "mock-project", "addresses", key)
+	}
+
+	if addr.Address == "" {
+		addr.Address = fmt.Sprintf("1.2.3.%d", ipCounter)
+		ipCounter++
+	}
+
```

Comment:
Need to add setting of default AddressType.
Need to add validation that address is not already reserved. 
https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/gce/gce_addresses_fakes.go#L65-L96

Reply:
Resolved